controllers/clustergroupupgrade_controller.go

```diff
@@ -1224,27 +1219,35 @@ func (r *ClusterGroupUpgradeReconciler) getNextNonCompliantPolicyForCluster(
             error/nil
 */
 func (r *ClusterGroupUpgradeReconciler) isUpgradeComplete(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) bool {
-	// If we are not at the last batch, the upgrade clearly didn't complete.
-	if clusterGroupUpgrade.Status.Status.CurrentBatch < len(clusterGroupUpgrade.Status.RemediationPlan) {
+	isBatchComplete, err := r.getNextRemediationPoliciesForBatch(ctx, clusterGroupUpgrade)
+	if err != nil {
 		return false
 	}
 
-	// Go through all the clusters in the current batch (which is also the last batch) and make sure
-	// they are either compliant with all the managedPolicies or they don't match them.
-	for _, batchClusterName := range clusterGroupUpgrade.Status.RemediationPlan[clusterGroupUpgrade.Status.Status.CurrentBatch-1] {
-		clusterProgress := clusterGroupUpgrade.Status.Status.CurrentBatchRemediationProgress[batchClusterName]
-		if clusterProgress.State == ranv1alpha1.Completed {
-			continue
-		}
-		nextNonCompliantPolicyIndex, err := r.getNextNonCompliantPolicyForCluster(ctx, clusterGroupUpgrade, batchClusterName, *clusterProgress.PolicyIndex)
-		if err != nil {
-			return false
-		}
-		if nextNonCompliantPolicyIndex < len(clusterGroupUpgrade.Status.ManagedPoliciesForUpgrade) {
-			return false
+	if isBatchComplete {
+		// Check previous batches
+		for i := 0; i < len(clusterGroupUpgrade.Status.RemediationPlan)-1; i++ {
+			for _, batchClusterName := range clusterGroupUpgrade.Status.RemediationPlan[i] {
+				clusterProgress := clusterGroupUpgrade.Status.Status.CurrentBatchRemediationProgress[batchClusterName]
+				if clusterProgress.State == ranv1alpha1.Completed {
+					continue
+				}
+				startIndex := 0
+				if clusterProgress.PolicyIndex != nil {
+					startIndex = *clusterProgress.PolicyIndex
+				}
+				nextNonCompliantPolicyIndex, err := r.getNextNonCompliantPolicyForCluster(ctx, clusterGroupUpgrade, batchClusterName, startIndex)
+				if err != nil {
```

Comment:
better syntax 

				if err != nil || nextNonCompliantPolicyIndex < len(clusterGroupUpgrade.Status.ManagedPoliciesForUpgrade)  {
					return false
				}

Reply:
Good point. Err should have been handled differently.